torchmultimodal/models/gpt.py

```diff
@@ -28,12 +28,221 @@ class TransformerLayerOutput(NamedTuple):
+        self.mm_decoder = mm_decoder
+        self.norm = nn.LayerNorm(d_model)
+        self.to_logit = nn.Linear(d_model, num_tokens, bias=False)
+        # This will give us equal probabilities after the soft max layer initially to avoid biasing
+        # towards any particular prediction category
+        self.to_logit.weight.data.copy_(torch.zeros(self.num_tokens, self.d_model))
+
+    def forward(
+        self,
+        in_modality: Optional[Tensor] = None,
+        out_modality: Optional[Tensor] = None,
+        in_pos_ids: Optional[Tensor] = None,
+        out_pos_ids: Optional[Tensor] = None,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        logits_mask: Optional[Tensor] = None,
+        use_cache: bool = False,
+        causal: bool = False,
+        right_shift: bool = False,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> MultimodalGPTOutput:
+        decoder_output = self.fwd(
+            in_modality=in_modality,
+            out_modality=out_modality,
+            in_pos_ids=in_pos_ids,
+            out_pos_ids=out_pos_ids,
+            attn_mask=attn_mask,
+            head_mask=head_mask,
+            use_cache=use_cache,
+            causal=causal,
+            right_shift=right_shift,
+            return_attn_weights=return_attn_weights,
+            return_hidden_states=return_hidden_states,
+        )
+
+        hidden_states = decoder_output.last_hidden_states
+        logits = self.head(hidden_states, logits_mask)
+
+        return MultimodalGPTOutput(decoder_output, logits)
+
+    def fwd(
+        self,
+        in_modality: Optional[Tensor] = None,
+        out_modality: Optional[Tensor] = None,
+        in_pos_ids: Optional[Tensor] = None,
+        out_pos_ids: Optional[Tensor] = None,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        use_cache: bool = False,
+        causal: bool = False,
+        right_shift: bool = False,
+        return_attn_weights: bool = False,
+        return_hidden_states: bool = False,
+    ) -> TransformerDecoderOutput:
+        # During training this method is used in the forward pass to decode input- and output- tokens.
+        # During generation this method is used for autoregressive decoding.
+        return self.mm_decoder(
+            in_modality=in_modality,
+            out_modality=out_modality,
+            in_pos_ids=in_pos_ids,
+            out_pos_ids=out_pos_ids,
+            attn_mask=attn_mask,
+            head_mask=head_mask,
+            use_cache=use_cache,
+            causal=causal,
+            right_shift=right_shift,
+            return_attn_weights=return_attn_weights,
+            return_hidden_states=return_hidden_states,
+        )
+
+    def head(
+        self, hidden_states: Tensor, logits_mask: Optional[Tensor] = None
+    ) -> Tensor:
+        out = self.norm(hidden_states)
```

Comment:
layernorm is usually not with the head (ex: VisionTransformer from torchvision [puts it with encoder](https://github.com/pytorch/vision/blob/cc0a83856c979f834132a5faf71a2eece0d32dc3/torchvision/models/vision_transformer.py#L157), and FLAVA has it separated, [image](https://github.com/facebookresearch/multimodal/blob/f9327dec04a5e8c824fba36e4ed52528e44541f6/torchmultimodal/models/flava/flava_image_encoder.py#L269) and [text](https://github.com/facebookresearch/multimodal/blob/f9327dec04a5e8c824fba36e4ed52528e44541f6/torchmultimodal/models/flava/flava_text_encoder.py#L218) encoder have a separate `Fp32LayerNorm` from the `Pooler` MLP head. I am not entirely sure if this is a convention, but maybe you could consider moving it out of `head` into `forward` or `fwd`.

Reply:
LayerNorm and logit layer are attached to the transformer decoder trunk during [training](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/gpt.py#L109-L114). However, they need to be [detached](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/gpt.py#L219-L226) from the trunk] when learning the key/value inner representation for the in-modality and then [attached](https://github.com/mugen-org/MUGEN_baseline/blob/783b5666186cae9e0517209925f0021db4026c19/lib/models/gpt/gpt.py#L227-L229) again when generating the next out-modality token. 

Grouping them together behind a method makes it flexible to attach/detach as needed.

The examples you cited are TransformerEncoder which may also be applicable for TransformerDecoder if the latter does not need to support AR decoding.

Reply:
ok I see, makes sense